### PR TITLE
Determine `Filter` unique keys efficiently

### PR DESCRIPTION
**Pull request by frankmcsherry**
Our `Filter` unique key determination would take exponential time (and produce exponential output) on queries like
```sql
create table test (a0 int, a1 int, a2 int, a3 int, a4 int, a5 int);

create materialized view temp as select distinct on (a0, a1, a2, a3, a4, a5) * from test;

select * from temp where a0 = a1 and a1 = a2 and a2 = a3 and a3 = a4 and a4 = a5;
```
The problem is that `temp` has as a unique key `[0, 1, 2, 3, 4, 5]` and the equivalences cause us to consider every possible substitution, which incudes all subsets of the columns. In fact, we only need as keys `[0], [1], [2], [3], [4], [5]`.

The new logic first reduces the input unique keys by equivalences, and then expands them back out. It takes time ~linear in the input and output keys (modulo some "looking up" in sorted column lists, which we could turn into binary search if needed). E.g. for the case above, it would go from the input key down to `[0]`, and then expand back out to each of the six columns, bypassing the exponential intermediate results.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aalexandrov** [comment]
Failing tests [in the Nightlies run](https://buildkite.com/materialize/nightlies/builds/6943#018e51b0-a394-4848-8b2a-2e176f47b9c6):

- Platform checks
- Legacy upgrade tests (last version from git)

I'm not seeing those in recent nightlies builds, so I'll have to investigate what is going on here before we can merge it.

File: src/expr/src/relation/mod.rs
Comment (aalexandrov) [inline src/expr/src/relation/mod.rs]: nit: this and the `input.dedup()` can be omitted as those are also done on the result-bound `keys`.

File: src/expr/src/relation/mod.rs
Comment (aalexandrov) [inline src/expr/src/relation/mod.rs:793]: nit: I wonder if we shouldn't try to do this on the `keys` bound to this result (so for all cases) as well.

**Review by aalexandrov — APPROVED** [review]
LGTM, but let's only merge this if the failing Nightlies pass on the rebased branch.

@frankmcsherry I've pushed some readability improvement suggestions [as a separate commit](https://github.com/MaterializeInc/materialize/pull/25708/commits/c7c2603fc8a096e6565f63188cee681559e1fbd6). LMK if you object to any of those, otherwise I will squash into the original commit before merging this.

File: src/expr/src/relation/mod.rs
Comment (frankmcsherry) [reply src/expr/src/relation/mod.rs:793]: I'm sorry I don't follow. It sounds like applying something more broadly, which seems great, but I don't know what you mean! :D

File: src/expr/src/relation/mod.rs
Comment (frankmcsherry) [inline src/expr/src/relation/mod.rs]: I didn't follow the reasoning for why this is not required, but if not it would be great to leave a comment with the reasoning. For example, why is it in sorted order already if for example we were to replace the first element of a key? Or perhaps we don't need sorted order (probably).

Comment (aalexandrov) [reply src/expr/src/relation/mod.rs]: I left [a comment here](https://github.com/MaterializeInc/materialize/pull/25708#discussion_r1528775168) but it seems that GitHub is hiding it now because I also removed the two lines in the proposed `fixup!` commit. The content was:

> nit: this and the input.dedup() can be omitted as those are also done on the result-bound keys.

This should also explain [my other comment above](https://github.com/MaterializeInc/materialize/pull/25708#discussion_r1528777440)

I mean that we can since we have

```rust
keys.sort();
keys.dedup();
```

at the end of `keys_with_input_keys`, we can possibly also do the 

```rust
for key_set in keys.iter_mut() {
    key_set.sort();
    key_set.dedup();
}
```

there and ensure that `key_set`s emitted from all cases in the 
```rust
let mut keys = match self {
    ...
}
```
branch appear in a sorted and deduped form.

Comment (aalexandrov) [reply src/expr/src/relation/mod.rs]: Beyond the scope of this PR, but: maybe we can introduce a `Key` struct that keeps its internal state in this canonical form (sorted, duplicate free). Then checking whether `k1: Key ⊆ k2: Key` should be doable in linear time.

**Comment by aalexandrov** [comment]
The RQG run with `EXPLAIN` reported several changed plans, and at least for three of those I see minor regressions (filter predicate conjuncts not being simplified).

@frankmcsherry I'm attaching [the traces or the reported queries (sadly not very small) against `v0.92.1` here](https://github.com/MaterializeInc/materialize/files/14650603/gh025708v0.9.zip) in case you want to dig deeper.

Given these changes, I suggest to merge this behind a feature flag (which can be enabled by default) in order to have a mechanism to quickly revert to the original behavior in case we hit an environment with a serious regression next week. If everything goes smoothly we can remove the feature flag approximately two weeks from now.

**Pull request merged**
